src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java

```diff
@@ -0,0 +1,617 @@
+
+    PathPlannerPath path = PathPlannerPath.fromPathPoints(Arrays.asList(new PathPoint(finalTransform.getTranslation(),
+        new RotationTarget(Constants.PLACEHOLDER_DOUBLE, finalHolonomicRotation))),
+        new PathConstraints(DriveConstants.SLOW_LINEAR_SPEED_METERS_PER_SEC,
+            DriveConstants.SLOW_LINEAR_ACCELERATION_METERS_PER_SEC_SQ, DriveConstants.SLOW_ANGULAR_SPEED_RAD_PER_SEC,
+            DriveConstants.SLOW_ANGULAR_ACCELERATION_RAD_PER_SEC_SQ),
+        null);
+    boolean generatedPath = true;
+    return Optional.of(path);
+  }
+
+  /**
+   * Driving inputs will get multiplied by the throttle value, so it should be in
+   * [0,1]
+   */
+  public void throttle(double throttleValue) {
+    throttleMultiplier = throttleValue;
+  }
+
+  public void offsetCurrentHeading(double offsetDegrees) {
+    targetHeadingDegrees = getHeadingDegrees() + offsetDegrees;
+  }
+
+  public void setZeroTargetHeading() {
+    targetHeadingDegrees = 0.0;
+  }
+
+  public void stopDriving() {
+    drive(0, 0, 0, true);
+  }
+
+  public Command stopDrivingCommand() {
+    return new InstantCommand(this::stopDriving, this);
+  }
+
+  public Command turnInPlace(double timeoutSec) {
+    return new RunCommand(
+        () -> {
+          rotateOrKeepHeading(0, 0, 0, true, -1);
+        })
+        .withTimeout(timeoutSec);
+  }
+
+  public void zeroFrontLeftAtCurrentPos() {
+    DriveCal.SWERVE_FRONT_LEFT_ANGULAR_OFFSET_RAD = frontLeft.getEncoderAbsPositionRad();
+    System.out.println(
+        "New Zero for Front Left Swerve: "
+            + DriveCal.SWERVE_FRONT_LEFT_ANGULAR_OFFSET_RAD);
+  }
+
+  public void zeroFrontRightAtCurrentPos() {
+    DriveCal.SWERVE_FRONT_RIGHT_ANGULAR_OFFSET_RAD = frontRight.getEncoderAbsPositionRad();
+    System.out.println(
+        "New Zero for Front Right Swerve: "
+            + DriveCal.SWERVE_FRONT_RIGHT_ANGULAR_OFFSET_RAD);
+  }
+
+  public void zeroRearLeftAtCurrentPos() {
+    DriveCal.SWERVE_BACK_LEFT_ANGULAR_OFFSET_RAD = rearLeft.getEncoderAbsPositionRad();
+    System.out.println(
+        "New Zero for Rear Left Swerve: "
+            + DriveCal.SWERVE_BACK_LEFT_ANGULAR_OFFSET_RAD);
+  }
+
+  public void zeroRearRightAtCurrentPos() {
+    DriveCal.SWERVE_BACK_RIGHT_ANGULAR_OFFSET_RAD = rearRight.getEncoderAbsPositionRad();
+    System.out.println(
+        "New Zero for Rear Right Swerve: "
+            + DriveCal.SWERVE_BACK_RIGHT_ANGULAR_OFFSET_RAD);
+  }
+
+  @Override
+  public void initSendable(SendableBuilder builder) {
+    super.initSendable(builder);
+    // addChild("X Controller", DriveCal.PATH_X_CONTROLLER);
+    // addChild("Y Controller", DriveCal.PATH_Y_CONTROLLER);
+    // addChild("Theta Controller", DriveCal.PATH_THETA_CONTROLLER);
+    addChild("Rotate to target controller", DriveCal.ROTATE_TO_TARGET_PID_CONTROLLER);
+    addChild("Front Right", frontRight);
+    addChild("Front Left", frontLeft);
+    addChild("Rear Right", rearRight);
+    addChild("Rear Left", rearLeft);
```

Comment:
Let's remember to try getting these to actually work this year

Reply:
*surely*